### PR TITLE
feat: split out lazy load package

### PR DESCRIPTION
**Pull request by craigbilner**
When `author-profile` went live we implemented lazy loading of the images at the package level. This was later refactored into `article-list` but actually is needed for `article` as well. This PR splits `LazyLoad` into it's own package and provides the ability for an article to also lazy load it's resources, currently related articles don't have high res images for example.

n.b. there's a slight behavioural difference in that browsers without `IntersectionObserver` now get back all the registered nodes as "observed" to make the API more consistent.

feat: create `lazy-load` package which solely handles observation of web nodes through the `IntersectionObserver`
refactor: remove lazy loading from `article-list`
test: remove lazy specific tests from `article-list`

### CODE REVIEW

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/82f4630b3f1ebdbecbc640752b24e35b64c957e9/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/eba589f0aa7bf4f0871f513610e94442a917f308/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/5d20ebdac60d8e65ff1a5199299519bb561574c0/index.html

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://s3-eu-west-1.amazonaws.com/times-components-snaps/c45d1eebdb96429f760519ded75b36f541a5f56f/index.html

File: packages/article-list/README.md [hunk 1]
```diff
@@ -24,10 +24,6 @@ unchanged.
 
 ## Lazy Loading
```
Comment (Belema) [inline packages/article-list/README.md]: Nothing left under this heading?

File: packages/article-list/__tests__/web/article-list-lazy.web.test.js [hunk 2]
```diff
@@ -1,3 +1,207 @@
...
+  };
+});
+
+afterEach(() => {
+  global.Intl = realIntl;
+  jest.restoreAllMocks();
+});
+
+beforeAll(() => jest.useRealTimers());
+afterAll(() => jest.useFakeTimers());
```
Comment (Belema) [inline packages/article-list/__tests__/web/article-list-lazy.web.test.js]: This sounds counter intuitive

File: packages/article-list/__tests__/web/article-list-lazy.web.test.js [hunk 3]
```diff
@@ -1,3 +1,207 @@
...
+
+    instance.cb(makeEntries(instance.nodes));
+  }
+
+  disconnect() {
+    return this;
+  }
+}
+
+afterEach(() => {
```
Comment (Belema) [inline packages/article-list/__tests__/web/article-list-lazy.web.test.js]: another `afterEach`?

File: packages/article-list/__tests__/web/article-list-lazy.web.test.js [hunk 4]
```diff
@@ -1,3 +1,207 @@
...
+
+const tests = [
+  {
+    name: "should render a good quality image if it is visible",
+    async test() {
+      window.IntersectionObserver = FakeIntersectionObserver;
+
+      const component = mount(<ArticleList {...articleListProps} />);
+
+      // prove the first image starts off as low quality
```
Comment (Belema) [inline packages/article-list/__tests__/web/article-list-lazy.web.test.js]: Should be another test?

File: packages/article-list/__tests__/web/__snapshots__/article-list.web.test.js.snap [hunk 5]
```diff
@@ -65,7 +65,7 @@ exports[`1. article list 1`] = `
           <div>
             <Card
               fadeImageIn={true}
-              highResSize={1440}
+              highResSize={null}
```
[the remark this answers is not in the document]
Comment (craigbilner) [reply packages/article-list/__tests__/web/__snapshots__/article-list.web.test.js.snap]: these tests are covered separately in the lazy test in `article-list` and then in `lazy-load` itself

File: packages/article-list/README.md (hunk 1 above)
Comment (craigbilner) [reply packages/article-list/README.md]: oops, ta

File: packages/lazy-load/README.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
+# Lazy Load
+
+This is implemented with
+[`IntersectionObserver`](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API)
+and therefore is web only.
+
+There are two props `rootMargin` and `threshold` which are as is in the docs
+above. Be sure to set a sensible threshold otherwise you may not get the lazy
+loading you were expecting. For example a threshold of 0 would mean no laziness.
```
Comment (Belema) [inline packages/lazy-load/README.md]: That's not what the documentation is saying.
> The default is 0 (meaning as soon as even one pixel is visible, the callback will be run).

File: packages/lazy-load/src/lazy-load.js [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+      } else if (
+        intersectionRatio < this.props.threshold &&
+        this.pending.has(target)
+      ) {
+        this.pending.delete(target);
+      }
+    });
+
+    if (this.pending.size) {
+      clearTimeout(this.pendingTimer);
```
Comment (Belema) [inline packages/lazy-load/src/lazy-load.js]: Why is this line inside of the `if (this.pending.size)`? Don't we want to always clear the timer?

File: packages/lazy-load/src/lazy-load.js [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+class LazyLoad extends Component {
+  constructor(props) {
+    super(props);
+
+    this.pending = new Set();
+    this.pendingTimer = null;
+    this.state = {
+      clientHasRendered: false,
+      isObserving: false,
```
Comment (Belema) [inline packages/lazy-load/src/lazy-load.js]: Why is `isObserving` part of the state? It never changes..

File: packages/article-list/__tests__/web/article-list-lazy.web.test.js (hunk 2 above)
Comment (craigbilner) [reply packages/article-list/__tests__/web/article-list-lazy.web.test.js]: ta, removed

File: packages/article-list/__tests__/web/article-list-lazy.web.test.js (hunk 3 above)
Comment (craigbilner) [reply packages/article-list/__tests__/web/article-list-lazy.web.test.js]: cool, ta

File: packages/lazy-load/README.md (hunk 6 above)
Comment (craigbilner) [reply packages/lazy-load/README.md]: :speaking_head: MDN docs are a little confusing in that they don't suggest the observed behaviour which is, if given a threshold of zero, all nodes will be "observed" immediately

File: packages/lazy-load/__tests__/web/lazy.web.test.js [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+const omitProps = new Set(["class", "className", "style"]);
+
+addSerializers(
+  expect,
+  enzymeRenderedSerializer(),
+  minimalise((value, key) => omitProps.has(key))
+);
+
+beforeAll(() => jest.useRealTimers());
+afterAll(() => jest.useFakeTimers());
```
Comment (Belema) [inline packages/lazy-load/__tests__/web/lazy.web.test.js]: Can these ones be removed as well?

File: packages/lazy-load/README.md (hunk 6 above)
Comment (Belema) [reply packages/lazy-load/README.md]: Ok, thanks

**Pull request merged**